src/replay/scheduler.zig

```diff
@@ -0,0 +1,294 @@
+                    if (try self.tryScheduleSome()) |err| {
+                        self.exit.store(true, .monotonic);
+                        self.failure = .{ .invalid_transaction = err };
+                    }
+                    return .pending;
+                } else {
+                    assert(self.batches.items.len == self.batches_finished);
+                    return .done;
+                }
+            },
+            .pending => return .pending,
+            .err => |err| return err,
+        }
+    }
+
+    fn tryScheduleSome(self: *TransactionScheduler) !?TransactionError {
+        while (self.batches.items.len > self.batches_started) {
+            const batch = self.batches.items[self.batches_started];
+            self.locks.lockStrict(self.allocator, batch.accounts) catch |e| {
+                switch (e) {
+                    error.LockFailed => if (self.batches_started == self.batches_finished) {
+                        return .AccountInUse;
+                    } else {
+                        break;
+                    },
+                    else => return e,
+                }
+            };
+            assert(try self.thread_pool.trySchedule(self.allocator, .{
+                .transactions = batch.transactions,
+                .results = &self.results,
+                .exit = &self.exit,
+            }));
+            self.batches_started += 1;
+        }
+        return null;
+    }
+};
+
+const ProcessBatchTask = struct {
+    transactions: []const ResolvedTransaction,
+    results: *Channel(?TransactionError),
+    exit: *std.atomic.Value(bool),
+
+    pub fn run(self: *ProcessBatchTask) !void {
+        const result = processBatch(self.transactions, self.exit);
+        if (result != null) self.exit.store(true, .monotonic);
+        try self.results.send(result);
+    }
+};
+
+test "TransactionScheduler: happy path" {
+    const allocator = std.testing.allocator;
+    const Transaction = sig.core.Transaction;
+    var rng = std.Random.DefaultPrng.init(123);
+
+    var thread_pool = ThreadPool.init(.{});
+    var scheduler = try TransactionScheduler.initCapacity(allocator, 10, &thread_pool);
+    defer scheduler.deinit();
+
+    const transactions = [_]Transaction{
+        try Transaction.initRandom(allocator, rng.random()),
+        try Transaction.initRandom(allocator, rng.random()),
+        try Transaction.initRandom(allocator, rng.random()),
+        try Transaction.initRandom(allocator, rng.random()),
+        try Transaction.initRandom(allocator, rng.random()),
+        try Transaction.initRandom(allocator, rng.random()),
+    };
+    defer for (transactions) |tx| tx.deinit(allocator);
+
+    {
+        const batch1 = try replay.resolve_lookup
+            .resolveBatch(allocator, undefined, transactions[0..3]);
+        errdefer batch1.deinit(allocator);
+
+        const batch2 = try replay.resolve_lookup
+            .resolveBatch(allocator, undefined, transactions[3..6]);
```

Comment:
Why is passing undefined here okay?

Reply:
Would be nice to encode this mode of operation with a dedicated noop provider, so as to trigger `error.PubkeyNotInIndex`, rather than illegal behavior.

Reply:
90d26d4c6967b51e99c28bfe79470f56c58e9a5b